src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java

```diff
@@ -0,0 +1,107 @@
+package org.springframework.data.elasticsearch.core;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.common.collect.MapBuilder;
+import org.elasticsearch.index.query.MoreLikeThisQueryBuilder;
+import org.elasticsearch.search.SearchHit;
+import org.springframework.data.elasticsearch.annotations.Document;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentEntity;
+import org.springframework.data.elasticsearch.core.mapping.ElasticsearchPersistentProperty;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Artur Konczak
+ * @author Mohsin Husen
+ * @author Mark Paluch
+ * @author Mark Janssen
+ * @author Sascha Woo
+ */
+public abstract class AbstractElasticsearchTemplate implements ElasticsearchOperations {
+
+	protected static String[] toArray(List<String> values) {
```

Comment:
Let's try to not expose `protected` utility methods as API. How about introducing a package-private utility class for Spring Data ES to ES API conversion. Alternatively, how about default methods on `Query` to retrieve fields, indices, types as array?